plugins/modules/ipatrust.py

```diff
@@ -0,0 +1,282 @@
+            range_type=dict(type="str", default="ipa-ad-trust", required=False, choices=[
+                            "ipa-ad-trust-posix", "ipa-ad-trust"]),
+            two_way=dict(type="bool", default=False, required=False),
+            external=dict(type="bool", default=False, required=False),
+        ),
+        supports_check_mode=True,
+    )
+
+    ansible_module._ansible_debug = True
+
+    # general
+    ipaadmin_principal = module_params_get(
+        ansible_module, "ipaadmin_principal")
+    ipaadmin_password = module_params_get(ansible_module, "ipaadmin_password")
+    realm = module_params_get(ansible_module, "realm")
+
+    # state
+    state = module_params_get(ansible_module, "state")
+
+    # trust
+    trust_type = module_params_get(ansible_module, "trust_type")
+    realm_admin = module_params_get(ansible_module, "admin")
+    realm_passwd = module_params_get(ansible_module, "password")
+    realm_server = module_params_get(ansible_module, "server")
+    trust_secret = module_params_get(ansible_module, "trust_secret")
+    base_id = module_params_get(ansible_module, "base_id")
+    range_size = module_params_get(ansible_module, "range_size")
+    range_type = module_params_get(ansible_module, "range_type")
+    bidirectional = module_params_get(ansible_module, "two_way")
+    external = module_params_get(ansible_module, "external")
+
+    changed = False
+    exit_args = {}
+    ccache_dir = None
+    ccache_name = None
+    try:
+        if not valid_creds(ansible_module, ipaadmin_principal):
+            ccache_dir, ccache_name = temp_kinit(
+                ipaadmin_principal, ipaadmin_password)
+        api_connect()
+        res_find = find_trust(ansible_module, realm)
+
+        if state == "absent" and res_find is not None:
+            del_trust(ansible_module, realm)
+            changed = True
+
+        if state == "present" and res_find is None:
+            exclusive = ["realm_admin", "realm_passwd"]
+            if vars()['trust_secret'] is not None:
+                for x in exclusive:
+                    if vars()[x] is not None:
+                        ansible_module.fail_json(
+                            msg="Argument '%s' is mutually exclusive with trust_secret" % (x))
+            elif vars()['realm_admin'] is None:
+                ansible_module.fail_json(
+                    msg="State present requires trust-secret or admin to be set")
+            elif vars()['realm_passwd'] is None:
+                ansible_module.fail_json(
+                    msg="State present with admin set requires password to be set")
+
+            args = gen_args(trust_type, realm_admin, realm_passwd, realm_server,
+                            trust_secret, base_id, range_size, range_type, bidirectional, external)
+
+            add_trust(ansible_module, realm, args)
+            changed = True
+
+    except Exception as e:
+        ansible_module.fail_json(msg='failed error :"%i"' % (int(e)))
```

Comment:
This does not work under IPA 4.4.0 with Python 2. I suggest using str(e), adapt accordingly.

Reply:
done